src/main/java/com/launchdarkly/openfeature/serverprovider/Provider.java

```diff
@@ -145,44 +147,66 @@ public void initialize(EvaluationContext evaluationContext) throws Exception {
-                    // reason we are stale or interrupted, so we want to emit an event each time.
-                    if (state != ProviderState.READY) {
-                        state = ProviderState.READY;
-                        emitProviderReady(ProviderEventDetails.builder().build());
+                switch (res.getState()) {
+                    // We will not re-enter INITIALIZING, but it is here to make the switch exhaustive.
+                    case INITIALIZING: {
                     }
+                    break;
+                    case INTERRUPTED: {
+                        synchronized (stateLock) {
+                            state = ProviderState.STALE;
+                        }
+
+                        var message = res.getLastError() != null ? res.getLastError().getMessage() : "encountered an unknown error";
+                        emitProviderStale(ProviderEventDetails.builder().message(message).build());
+                    }
+                    break;
+                    case VALID: {
+                        boolean emit = false;
+                        synchronized (stateLock) {
+                            // If we are ready, then we don't want to emit it again. Other conditions we may be updating the
+                            // reason we are stale or interrupted, so we want to emit an event each time.
+                            if (state != ProviderState.READY) {
+                                state = ProviderState.READY;
+                                emit = true;
+                            }
+                        }
+                        if(emit) {
+                            emitProviderReady(ProviderEventDetails.builder().build());
+                        }
+                    }
+                    break;
+                    case OFF: {
+                        synchronized (stateLock) {
+                            // Currently there is not a shutdown state.
+                            // Our client/provider cannot be restarted, so we just go to error.
+                            state = ProviderState.ERROR;
+                        }
+                        emitProviderError(ProviderEventDetails.builder().message("Provider shutdown").build());
+                    }
+                    stateLock.notify();
                 }
-                break;
-                case OFF: {
-                    // Currently there is not a shutdown state.
-                    // Our client/provider cannot be restarted, so we just go to error.
-                    state = ProviderState.ERROR;
-                    emitProviderError(ProviderEventDetails.builder().message("Provider shutdown").build());
-                }
-            }
         });
         if (state == ProviderState.READY) {
             return;
         }
 
-        boolean initialized = client.getDataSourceStatusProvider().waitFor(DataSourceStatusProvider.State.VALID,
```

Comment:
There were a couple problems here. The `waitFor` method just not handle FOREVER, and so was doing some invalid math.

Also it really needs to check for either OFF or VALID, which this API doesn't really have an ergonomic way to do.

So instead we use synchronized blocks with wait. Whenever the state changes the checking code wakes up and checks if initialization is complete.